### PR TITLE
Upgrade to 0.13.2, merge recent LLM changes

### PR DESCRIPTION
**Pull request by abhi-mosaic**
This PR upgrades to Composer 0.13.2 and loops in all pending changes from the last few weeks of LLM runs.

Going to do some LLM runs on this branch to test and send screenshots.

### CODE REVIEW

File: examples/llm/main.py [hunk 1]
```diff
@@ -86,7 +86,8 @@ def main(cfg):
     )
 
     reproducibility.seed_all(cfg.seed)
-    dist.initialize_dist(get_device(None))
+    dist.initialize_dist(get_device(None),
+                         timeout=cfg.get('dist_timeout', 1500.0))
```
Comment (dakinggg) [inline examples/llm/main.py:90]: for completeness, can you also pass this `dist_timeout` to trainer?

File: examples/llm/requirements.txt [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
+mosaicml-streaming==0.3.0
 flash-attn==0.2.8
 mosaicml-cli>=0.2.32,<1
 # need a newer version of triton
 triton==2.0.0.dev20221202
-transformers==4.25.1
+transformers==4.26.1
 omegaconf==2.2.3
-wandb==0.13.4
+wandb==0.13.6
```
Comment (dakinggg) [inline examples/llm/requirements.txt:11]: any particular reason for changing this one? and did we have some issues upgrading it in the past?

**Review by dakinggg — APPROVED** [review]
LGTM once you've gotten a couple runs in successfully.
Request for manual testing: 
- a mosaicgpt run with icl
- a gptneo run with icl
- run through the bert readme
- run through at least the resnet imagenet readme (we can do the rest before release)

**Comment by dakinggg** [comment]
Also, you'll need to bump mosaicml versions in the other subfolders to increase the top level requirement

File: examples/common/optim/lion.py
Comment (vchiley) [inline examples/common/optim/lion.py:1]: things are converging using this file. it is the gold standard. DO NOT MODIFY :)

File: examples/llm/icl_eval/evaluate_model.py [hunk 3]
```diff
@@ -21,13 +23,20 @@
     cli_cfg = om.from_cli(args_list)
     cfg = DictConfig(om.merge(yaml_cfg, cli_cfg))
 
+    reproducibility.seed_all(cfg.get('seed', 1234))
```
Comment (vchiley) [inline examples/llm/icl_eval/evaluate_model.py:26]: i like that we had `17` going

File: examples/llm/main.py (hunk 1 above)
Comment (vchiley) [inline examples/llm/main.py:90]: 1500 is a pretty random default (25 min)
300 is used somewhere in composer (5 min)
1800 is the torch dist default time out (30 min)

File: examples/llm/requirements.txt (hunk 2 above)
Comment (abhi-mosaic) [reply examples/llm/requirements.txt:11]: I see this more recent one being used by BERT and want to keep them in sync

File: examples/llm/src/models/layers/gpt_blocks.py [hunk 4]
```diff
@@ -36,9 +38,13 @@ def __init__(self,
         super().__init__()
         if cfg.get('alibi', False):
             assert cfg.attn_impl == 'triton' or cfg.attn_impl == 'torch', 'Only triton kernel or torch supports alibi'
-        self.ln_1 = nn.LayerNorm(cfg.d_model, device=device)
+
+        layernorm_class = nn.LayerNorm if not cfg.get(
+            'low_precision_layernorm', False) else LPLayerNorm
```
Comment (vchiley) [inline examples/llm/src/models/layers/gpt_blocks.py:43]:
```suggestion
        layernorm_class = LPLayerNorm if cfg.get(
            'low_precision_layernorm', False) else nn.LayerNorm
```

nit: avoids the `not`

File: examples/llm/src/models/mosaic_gpt.py [hunk 5]
```diff
@@ -38,6 +40,9 @@ def __init__(self, cfg: DictConfig):
         else:
             raise ValueError(f'Unknown attn_impl={cfg.attn_impl}')
 
+        layernorm_class = nn.LayerNorm if not cfg.get(
+            'low_precision_layernorm', False) else LPLayerNorm
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:44]:
```suggestion
        layernorm_class = LPLayerNorm if cfg.get(
            'low_precision_layernorm', False) else nn.LayerNorm
```

same nit

**Review by vchiley — APPROVED** [review]
LGTM, left a few nits

File: examples/bert/src/hf_bert.py [hunk 6]
```diff
@@ -217,8 +217,9 @@ def create_hf_bert_classification(
     else:
         # Metrics for a classification model
         metrics = [
-            Accuracy(),
-            MatthewsCorrCoef(num_classes=model.config.num_labels)
+            MulticlassAccuracy(num_classes=num_labels, average='micro'),
+            MatthewsCorrCoef(task='multiclass',
+                             num_classes=model.config.num_labels)
```
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply examples/bert/src/hf_bert.py:222]: I think this is a @dakinggg question, but there was some upgrade of torchmetrics inside Composer, and potentially some upgrades of our metric implementations, that necessitated this.

[the remark this answers is not in the document]
Comment (dakinggg) [reply examples/bert/src/hf_bert.py:222]: yeah, torchmetrics had some breaking changes between 0.9.3 and 0.11.3, and we did that upgrade in the latest composer.

File: examples/llm/main.py (hunk 1 above)
Comment (abhi-mosaic) [reply examples/llm/main.py]: done

**Comment by abhi-mosaic** [comment]
CI passing, just running through some test runs now...

**Comment by abhi-mosaic** [comment]
Successfully ran MosaicGPT-1.3B and OPT-2.7B both with ICL eval, and BERT pretraining, all is well!

**Pull request merged**